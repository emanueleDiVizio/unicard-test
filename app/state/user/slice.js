import {createAsyncThunk, createSlice} from '@reduxjs/toolkit';
import {userAPI} from '../../api';

const getUsers = createAsyncThunk(
  'user/getUsers',
  async (_, {rejectWithValue, getState}) => {
    const response = await userAPI.getUsers();
    if (response.ok) {
      return response.data;
    } else {
      return rejectWithValue(response.problem);
    }
  },
);

export const slice = createSlice({
  name: 'user',
  initialState: {
    users: [],
    isLoading: false,
    error: null,
  },
  reducers: {},
  extraReducers: {
    [getUsers.pending]: (state, action) => {
      if (!state.isLoading) {
        state.isLoading = true;
      }
    },
    [getUsers.fulfilled]: (state, action) => {
      if (state.isLoading) {
        state.isLoading = false;
        state.users = action.payload;
      }
    },
    [getUsers.rejected]: (state, action) => {
      if (state.isLoading) {
        state.isLoading = false;
        if (action.payload) {
          state.error = action.payload.errorMessage;
        } else {
          state.error = action.error;
        }
      }
    },
  },
});

export const thunks = {getUsers};
