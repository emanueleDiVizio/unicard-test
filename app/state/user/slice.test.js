import {slice, thunks} from './slice';

describe('state/user/slice', () => {
  describe('pending', () => {
    it('updates loading state if not loading', () => {
      expect(
        slice.reducer({isLoading: false}, thunks.getUsers.pending()),
      ).toMatchSnapshot();
    });
    it('does nothing if loading', () => {
      expect(
        slice.reducer({isLoading: true}, thunks.getUsers.pending()),
      ).toMatchSnapshot();
    });
  });
  describe('fulfilled', () => {
    it('stores new data if it was loading', () => {
      expect(
        slice.reducer(
          {isLoading: true},
          thunks.getUsers.fulfilled([{name: 'mockName'}]),
        ),
      ).toMatchSnapshot();
    });
    it('does nothing if it was not loading', () => {
      expect(
        slice.reducer(
          {isLoading: false},
          thunks.getUsers.fulfilled([{name: 'mockName'}]),
        ),
      ).toMatchSnapshot();
    });
  });
  describe('rejected', () => {
    it('does nothing if it was not loading', () => {
      expect(
        slice.reducer({isLoading: false}, thunks.getUsers.rejected()),
      ).toMatchSnapshot();
    });
    it('registers error if was present and request errored gracefully', () => {
      expect(
        slice.reducer({isLoading: true}, thunks.getUsers.rejected('PROBLEM')),
      ).toMatchSnapshot();
    });

    it('registers error if was present and request thrown', () => {
      expect(
        slice.reducer({isLoading: true}, thunks.getUsers.rejected()),
      ).toMatchSnapshot();
    });
  });
});
