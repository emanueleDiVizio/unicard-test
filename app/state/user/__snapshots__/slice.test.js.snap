// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state/user/slice fulfilled does nothing if it was not loading 1`] = `
Object {
  "isLoading": false,
}
`;

exports[`state/user/slice fulfilled stores new data if it was loading 1`] = `
Object {
  "isLoading": false,
  "users": Array [
    Object {
      "name": "mockName",
    },
  ],
}
`;

exports[`state/user/slice pending does nothing if loading 1`] = `
Object {
  "isLoading": true,
}
`;

exports[`state/user/slice pending updates loading state if not loading 1`] = `
Object {
  "isLoading": true,
}
`;

exports[`state/user/slice rejected does nothing if it was not loading 1`] = `
Object {
  "isLoading": false,
}
`;

exports[`state/user/slice rejected registers error if was present and request errored gracefully 1`] = `
Object {
  "error": Object {
    "message": "PROBLEM",
  },
  "isLoading": false,
}
`;

exports[`state/user/slice rejected registers error if was present and request thrown 1`] = `
Object {
  "error": Object {
    "message": "Rejected",
  },
  "isLoading": false,
}
`;
